File: webview/android/app/src/main/java/com/awesomeproject/FileServer.java
package com.awesomeproject;

import android.content.Context;
import android.webkit.MimeTypeMap;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;

import fi.iki.elonen.NanoHTTPD;

public class FileServer extends NanoHTTPD {
  private final Context context;

  public FileServer(Context context, int port) {
    super(port);
    this.context = context;
  }

  @Override
  public Response serve(IHTTPSession session) {
    String uri = session.getUri();
    try {
      InputStream inputStream = context.getAssets().open(uri.substring(1));
      return newChunkedResponse(Response.Status.OK, getMimeType(uri), inputStream);
    } catch (IOException e) {
      String str = new String();
      try {
        for (int i = 0; i < context.getAssets().list("").length; i++) {
          str += context.getAssets().list("")[i] + "    ";
        }
      } catch (IOException err) {
        return newFixedLengthResponse(Response.Status.NOT_FOUND, NanoHTTPD.MIME_PLAINTEXT, "File not found");
      }
      return newFixedLengthResponse(Response.Status.NOT_FOUND, NanoHTTPD.MIME_PLAINTEXT,
          "File not found." + str + ";" + context.getFilesDir());
    }
  }

  private String getMimeType(String filePath) {
    String extension = MimeTypeMap.getFileExtensionFromUrl(filePath);
    return MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension);
  }
}

File: webview/android/app/src/main/java/com/awesomeproject/FileServer.kt
package com.awesomeproject

import android.content.Context
import android.webkit.MimeTypeMap
import fi.iki.elonen.NanoHTTPD
import java.io.IOException

class FileServer(private val context: Context, port: Int) : NanoHTTPD(port) {
    override fun serve(session: IHTTPSession): Response {
        val uri = session.uri
        return try {
            val inputStream = context.assets.open(uri.substring(1))
            newChunkedResponse(Response.Status.OK, getMimeType(uri), inputStream)
        } catch (e: IOException) {
            val str = StringBuilder()
            try {
                for (i in context.assets.list("")!!.indices) {
                    str.append(context.assets.list("")!![i]).append("    ")
                }
            } catch (err: IOException) {
                return newFixedLengthResponse(
                        Response.Status.NOT_FOUND,
                        NanoHTTPD.MIME_PLAINTEXT,
                        "File not found"
                )
            }
            newFixedLengthResponse(
                    Response.Status.NOT_FOUND,
                    NanoHTTPD.MIME_PLAINTEXT,
                    "File not found.$str;${context.filesDir}"
            )
        }
    }

    private fun getMimeType(filePath: String): String? {
        val extension = MimeTypeMap.getFileExtensionFromUrl(filePath)
        return MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension)
    }
}
